bin/reth-benchmark/src/authenticated_transport.rs

```diff
@@ -0,0 +1,197 @@
+//! This contains an authenticated rpc transport that can be used to send engine API newPayload
+//! requests.
+
+use std::sync::Arc;
+
+use alloy_json_rpc::{RequestPacket, ResponsePacket};
+use alloy_pubsub::{PubSubConnect, PubSubFrontend};
+use alloy_rpc_types_engine::{Claims, JwtSecret};
+use alloy_transport::{Authorization, TransportError, TransportFut};
+use alloy_transport_http::{reqwest::Url, Http, ReqwestTransport};
+use alloy_transport_ipc::IpcConnect;
+use alloy_transport_ws::WsConnect;
+use reqwest::header::HeaderValue;
+use std::task::{Context, Poll};
+use tokio::sync::RwLock;
+use tower::Service;
+
+/// An enum representing the different transports that can be used to connect to a runtime.
+/// Only meant to be used internally by [AuthenticatedTransport].
+#[derive(Clone, Debug)]
+pub enum InnerTransport {
+    /// HTTP transport
+    Http(ReqwestTransport),
+    /// WebSocket transport
+    Ws(PubSubFrontend),
+    /// IPC transport
+    Ipc(PubSubFrontend),
+}
+
+#[derive(Debug, Clone)]
+pub struct AuthenticatedTransport {
+    /// The inner actual transport used.
+    inner: Arc<RwLock<InnerTransport>>,
+}
+
+/// An error that can occur when creating an authenticated transport.
+#[derive(Debug, thiserror::Error)]
+pub enum AuthenticatedTransportError {
+    /// The URL is invalid.
+    #[error("The URL is invalid")]
+    InvalidUrl,
+    /// Failed to lock transport
+    #[error("Failed to lock transport")]
+    LockFailed,
+    /// The JWT is invalid.
+    #[error("The JWT is invalid: {0}")]
+    InvalidJwt(String),
+    /// The transport failed to connect.
+    #[error("The transport failed to connect to {1}, transport error: {0}")]
+    TransportError(TransportError, String),
+    /// The http client could not be built.
+    #[error("The http client could not be built")]
+    HttpConstructionError(reqwest::Error),
+    /// The scheme is invalid.
+    #[error("The URL scheme is invalid: {0}")]
+    BadScheme(String),
+}
+
+impl AuthenticatedTransport {
+    /// Create a new builder with the given URL.
+    pub async fn connect(url: Url, jwt: JwtSecret) -> Result<Self, AuthenticatedTransportError> {
+        match url.scheme() {
+            "http" | "https" => Self::connect_http(url, jwt).await,
+            "ws" | "wss" => Self::connect_ws(url, jwt).await,
+            "file" => Self::connect_ipc(url).await,
+            _ => Err(AuthenticatedTransportError::BadScheme(url.scheme().to_string())),
+        }
```

Comment:
I thought you could use connect_builtin here instead from Alloy? And we should have a nicer way to https & JWT/AUTH headers?

Reply:
> I thought you could use connect_builtin here instead from Alloy? And we should have a nicer way to https & JWT/AUTH headers?

right, `connect_builtin` should work, still need to add the headers first, but we should definitely upstream some jwt client features in alloy. then this could be 2 lines total in `rpc.rs`